components/ntp_background_images/browser/ntp_background_images_service.cc

```diff
@@ -6,82 +6,394 @@
 #include "brave/components/ntp_background_images/browser/ntp_background_images_data.h"
 #include "brave/components/ntp_background_images/browser/ntp_background_images_source.h"
-#include "brave/components/ntp_background_images/browser/regional_component_data.h"
+#include "brave/components/ntp_background_images/browser/ntp_background_images_utils.h"
+#include "brave/components/ntp_background_images/browser/sponsored_images_component_data.h"
 #include "brave/components/ntp_background_images/browser/switches.h"
+#include "brave/components/ntp_background_images/browser/url_constants.h"
+#include "brave/components/ntp_background_images/common/pref_names.h"
+#include "components/component_updater/component_updater_service.h"
+#include "components/prefs/pref_service.h"
 
 namespace ntp_background_images {
 
 namespace {
 
-constexpr char kPhotoJsonFilename[] = "photo.json";
+constexpr char kNTPSIManifestFile[] = "photo.json";
+constexpr char kNTPSRManifestFile[] = "data.json";
+constexpr char kNTPSRMappingTableFile[] = "mapping-table.json";
 
-std::string ReadPhotosManifest(const base::FilePath& photos_manifest_path) {
+constexpr char kNTPSRMappingTableComponentPublicKey[] = "MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAp7IWv7wzH/KLrxx7BKWOIIUMDylQNzxwM5Fig2WHc16BoMW9Kaya/g17Bpfp0YIvxdcmDBcB9kFALqQLxi1WQfa9d7YxqcmAGUKo407RMwEa6dQVkIPMFz2ZPGSfFgr526gYOqWh3Q4h8oN94qxBLgFyT25SMK5zQDGyq96ntME4MQRNwpDBUv7DDK7Npwe9iE8cBgzYTvf0taAFn2ZZi1RhS0RzpdynucpKosnc0sVBLTXy+HDvnMr+77T48zM0YmpjIh8Qmrp9CNbKzZUsZzNfnHpL9IZnjwQ51EOYdPGX2r1obChVZN19HzpK5scZEMRKoCMfCepWpEkMSIoPzQIDAQAB";  // NOLINT
+constexpr char kNTPSRMappingTableComponentID[] =
+    "heplpbhjcbmiibdlchlanmdenffpiibo";
+constexpr char kNTPSRMappingTableComponentName[] =
+    "NTP Super Referral mapping table";
+
+std::string GetMappingTableData(const base::FilePath& installed_dir) {
   std::string contents;
-  bool success = base::ReadFileToString(photos_manifest_path, &contents);
-  if (!success || contents.empty()) {
-    DVLOG(2) << "ReadPhotosManifest: cannot "
-             << "read photo.json file " << photos_manifest_path;
-  }
+  const auto json_path = installed_dir.AppendASCII(kNTPSRMappingTableFile);
+  base::ReadFileToString(json_path, &contents);
   return contents;
 }
 
-NTPBackgroundImagesData* GetDemoWallpaper() {
-  static auto demo = std::make_unique<NTPBackgroundImagesData>();
-  demo->url_prefix = "chrome://newtab/ntp-dummy-brandedwallpaper/";
-  demo->backgrounds = {
-      { base::FilePath(FILE_PATH_LITERAL("wallpaper1.jpg")), { 3988, 2049 } },
-      { base::FilePath(FILE_PATH_LITERAL("wallpaper2.jpg")), { 5233, 3464 } },
-      { base::FilePath(FILE_PATH_LITERAL("wallpaper3.jpg")), {  0, 0 } },
-  };
-  demo->logo_alt_text = "Technikke: For music lovers.";
-  demo->logo_company_name = "Technikke";
-  demo->logo_destination_url = "https://brave.com";
-  return demo.get();
+void CacheSuperReferralData(const std::string& data_json,
```

Comment:
I'm not understanding what's happening here. What are we caching and why?

Reply:
Caching will be improved in the f/u.